Remove limitation on TLSv1 in FIPS mode

diff --git a/src/options.c b/src/options.c
index 183d4d4..d2f7d07 100644
--- a/src/options.c
+++ b/src/options.c
@@ -1682,14 +1682,6 @@ static char *parse_service_option(CMD cmd, SERVICE_OPTIONS *section,
             return "Incorrect version of SSL protocol";
         return NULL; /* OK */
     case CMD_END:
-#ifdef USE_FIPS
-        if(new_global_options.option.fips &&
-                ((section->option.client &&
-                    section->client_method!=(SSL_METHOD *)TLSv1_client_method()) ||
-                (!section->option.client &&
-                    section->server_method!=(SSL_METHOD *)TLSv1_server_method())))
-            return "'sslVersion = TLSv1' is required in FIPS mode";
-#endif /* USE_FIPS */
         break;
     case CMD_FREE:
         break;
